Use stable metric for const eval limit instead of current terminator-based logic

bryangarza opened the pull request:
This patch adds a `MirPass` that inserts a new MIR instruction `ConstEvalCounter` to any loops and function calls in the CFG. This instruction is used during Const Eval to count against the `const_eval_limit`, and emit the `StepLimitReached` error, replacing the current logic which uses Terminators only.

The new method of counting loops and function calls should be more stable across compiler versions (i.e., not cause crates that compiled successfully before, to no longer compile when changes to the MIR generation/optimization are made).

Also see: #103877

bryangarza committed:
Create stable metric to measure long computation in Const Eval

This patch adds a `MirPass` that tracks the number of back-edges and
function calls in the CFG, adds a new MIR instruction to increment a
counter every time they are encountered during Const Eval, and emit a
warning if a configured limit is breached.

bryangarza commented:
r? @pnkfelix

compiler-errors commented on compiler/rustc_mir_transform/src/ctfe_limit.rs:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        let mut dominators = Vec::new();
+        for idom in 0..body.basic_blocks.len() {
+            let mut nodes = Vec::new();
+            for inode in 0..body.basic_blocks.len() {
+                let dom = BasicBlock::from_usize(idom);
+                let node = BasicBlock::from_usize(inode);
```
```suggestion
        for (dom, _) in body.basic_blocks.iter_enumerated() {
            let mut nodes = Vec::new();
            for (node, _) in body.basic_blocks.iter_enumerated() {
```

You could probably use more iterator logic to avoid having to do these in `for` loops at all, just some `map()` + `collect()` steps.

compiler-errors commented on compiler/rustc_mir_transform/src/ctfe_limit.rs:
```diff
@@ -0,0 +1,92 @@
+        }
+        */
+
+        for (index, basic_block) in body.basic_blocks_mut().iter_mut().enumerate() {
```
You could use `iter_enumerated_mut` here, and also you could instead collect `dominators` into an `IndexVec` indexed by `BasicBlock` to avoid the `as_usize` below.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3796286551/jobs/6456239211) [(plain)](https://github.com/rust-lang/rust/commit/02196486c1b315a916e42f33f8ed45dc36f7f3bb/checks/10339957251/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0004]: non-exhaustive patterns: `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
   --> src/base.rs:515:11
    |
515 |     match &stmt.kind {
    |           ^^^^^^^^^^ pattern `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
note: `rustc_middle::mir::StatementKind<'_>` defined here
   --> /checkout/compiler/rustc_middle/src/mir/syntax.rs:361:5
    |
242 | pub enum StatementKind<'tcx> {
242 | pub enum StatementKind<'tcx> {
    | ----------------------------
...
361 |     ConstEvalCounter,
    |     ^^^^^^^^^^^^^^^^ not covered
    = note: the matched value is of type `&rustc_middle::mir::StatementKind<'_>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
822 ~         },
822 ~         },
823 ~         &rustc_middle::mir::StatementKind::ConstEvalCounter => todo!(),

error[E0004]: non-exhaustive patterns: `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
   --> src/constant.rs:481:27
    |
    |
481 |                     match &stmt.kind {
    |                           ^^^^^^^^^^ pattern `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
note: `rustc_middle::mir::StatementKind<'_>` defined here
   --> /checkout/compiler/rustc_middle/src/mir/syntax.rs:361:5
    |
242 | pub enum StatementKind<'tcx> {
242 | pub enum StatementKind<'tcx> {
    | ----------------------------
...
361 |     ConstEvalCounter,
    |     ^^^^^^^^^^^^^^^^ not covered
    = note: the matched value is of type `&rustc_middle::mir::StatementKind<'_>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
533 ~                         | StatementKind::Nop => {}
533 ~                         | StatementKind::Nop => {}
534 +                         &rustc_middle::mir::StatementKind::ConstEvalCounter => todo!()

For more information about this error, try `rustc --explain E0004`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:02:11
```

</details>

bryangarza committed:
Clean up CtfeLimit MirPass

bryangarza committed:
Remove debugging-related code

bryangarza commented:
Remaining:
- [ ] use `checked_add(..)` for incrementing counter
- [ ] Make use of `StepLimitReached` error and remove terminator-based logic from https://github.com/rust-lang/rust/blame/master/compiler/rustc_const_eval/src/const_eval/machine.rs#L564-L576

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3798183486/jobs/6459698199) [(plain)](https://github.com/rust-lang/rust/commit/8f2092464730e0cd7648cb093abd28e0a1401ea2/checks/10344192478/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0004]: non-exhaustive patterns: `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
   --> src/base.rs:515:11
    |
515 |     match &stmt.kind {
    |           ^^^^^^^^^^ pattern `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
note: `rustc_middle::mir::StatementKind<'_>` defined here
   --> /checkout/compiler/rustc_middle/src/mir/syntax.rs:361:5
    |
242 | pub enum StatementKind<'tcx> {
242 | pub enum StatementKind<'tcx> {
    | ----------------------------
...
361 |     ConstEvalCounter,
    |     ^^^^^^^^^^^^^^^^ not covered
    = note: the matched value is of type `&rustc_middle::mir::StatementKind<'_>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
822 ~         },
822 ~         },
823 ~         &rustc_middle::mir::StatementKind::ConstEvalCounter => todo!(),

error[E0004]: non-exhaustive patterns: `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
   --> src/constant.rs:481:27
    |
    |
481 |                     match &stmt.kind {
    |                           ^^^^^^^^^^ pattern `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
note: `rustc_middle::mir::StatementKind<'_>` defined here
   --> /checkout/compiler/rustc_middle/src/mir/syntax.rs:361:5
    |
242 | pub enum StatementKind<'tcx> {
242 | pub enum StatementKind<'tcx> {
    | ----------------------------
...
361 |     ConstEvalCounter,
    |     ^^^^^^^^^^^^^^^^ not covered
    = note: the matched value is of type `&rustc_middle::mir::StatementKind<'_>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
533 ~                         | StatementKind::Nop => {}
533 ~                         | StatementKind::Nop => {}
534 +                         &rustc_middle::mir::StatementKind::ConstEvalCounter => todo!()

For more information about this error, try `rustc --explain E0004`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:02:12
```

</details>

bryangarza committed:
Change code to use map insead of for-loop

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3798376427/jobs/6460037462) [(plain)](https://github.com/rust-lang/rust/commit/ea6d20e45eea835b5d6eca276420ff8fe3391b63/checks/10344600299/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0004]: non-exhaustive patterns: `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
   --> src/base.rs:515:11
    |
515 |     match &stmt.kind {
    |           ^^^^^^^^^^ pattern `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
note: `rustc_middle::mir::StatementKind<'_>` defined here
   --> /checkout/compiler/rustc_middle/src/mir/syntax.rs:361:5
    |
242 | pub enum StatementKind<'tcx> {
242 | pub enum StatementKind<'tcx> {
    | ----------------------------
...
361 |     ConstEvalCounter,
    |     ^^^^^^^^^^^^^^^^ not covered
    = note: the matched value is of type `&rustc_middle::mir::StatementKind<'_>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
822 ~         },
822 ~         },
823 ~         &rustc_middle::mir::StatementKind::ConstEvalCounter => todo!(),

error[E0004]: non-exhaustive patterns: `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
   --> src/constant.rs:481:27
    |
    |
481 |                     match &stmt.kind {
    |                           ^^^^^^^^^^ pattern `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
note: `rustc_middle::mir::StatementKind<'_>` defined here
   --> /checkout/compiler/rustc_middle/src/mir/syntax.rs:361:5
    |
242 | pub enum StatementKind<'tcx> {
242 | pub enum StatementKind<'tcx> {
    | ----------------------------
...
361 |     ConstEvalCounter,
    |     ^^^^^^^^^^^^^^^^ not covered
    = note: the matched value is of type `&rustc_middle::mir::StatementKind<'_>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
533 ~                         | StatementKind::Nop => {}
533 ~                         | StatementKind::Nop => {}
534 +                         &rustc_middle::mir::StatementKind::ConstEvalCounter => todo!()

For more information about this error, try `rustc --explain E0004`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:02:48
```

</details>

bryangarza committed:
Abstract out has_back_edge fn

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3802726221/jobs/6468476420) [(plain)](https://github.com/rust-lang/rust/commit/e0981a5c21ef64acf4f02e8fbab8eff5f6803098/checks/10355003307/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0004]: non-exhaustive patterns: `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
   --> src/base.rs:515:11
    |
515 |     match &stmt.kind {
    |           ^^^^^^^^^^ pattern `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
note: `rustc_middle::mir::StatementKind<'_>` defined here
   --> /checkout/compiler/rustc_middle/src/mir/syntax.rs:361:5
    |
242 | pub enum StatementKind<'tcx> {
242 | pub enum StatementKind<'tcx> {
    | ----------------------------
...
361 |     ConstEvalCounter,
    |     ^^^^^^^^^^^^^^^^ not covered
    = note: the matched value is of type `&rustc_middle::mir::StatementKind<'_>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
822 ~         },
822 ~         },
823 ~         &rustc_middle::mir::StatementKind::ConstEvalCounter => todo!(),

error[E0004]: non-exhaustive patterns: `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
   --> src/constant.rs:481:27
    |
    |
481 |                     match &stmt.kind {
    |                           ^^^^^^^^^^ pattern `&rustc_middle::mir::StatementKind::ConstEvalCounter` not covered
note: `rustc_middle::mir::StatementKind<'_>` defined here
   --> /checkout/compiler/rustc_middle/src/mir/syntax.rs:361:5
    |
242 | pub enum StatementKind<'tcx> {
242 | pub enum StatementKind<'tcx> {
    | ----------------------------
...
361 |     ConstEvalCounter,
    |     ^^^^^^^^^^^^^^^^ not covered
    = note: the matched value is of type `&rustc_middle::mir::StatementKind<'_>`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
533 ~                         | StatementKind::Nop => {}
533 ~                         | StatementKind::Nop => {}
534 +                         &rustc_middle::mir::StatementKind::ConstEvalCounter => todo!()

For more information about this error, try `rustc --explain E0004`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:02:20
```

</details>

bryangarza committed:
Replace terminator-based const eval limit

- Remove logic that limits const eval based on terminators, and use the
  stable metric instead (back edges + fn calls)
- Add unstable flag `tiny-const-eval-limit` to add UI tests that do not
  have to go up to the regular 2M step limit

bryangarza commented:
@rustbot ready

bryangarza committed:
Update codegen cranelift for ConstEvalCounter

bryangarza committed:
Update Clippy for ConstEvalCounter

bryangarza committed:
Move CtfeLimit to mir_const's set of passes

bryangarza committed:
Add back Machine::before_terminator(...) method

Added it back because it's used by Miri, but in the compiler itself, it
will not do anything (just return `Ok(())`.

bryangarza committed:
Revert "Move CtfeLimit to mir_const's set of passes"

This reverts commit 332542a92223b2800ed372d2d461921147f29477.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3804242786/jobs/6471310660) [(plain)](https://github.com/rust-lang/rust/commit/a4d1eb7b5a3675f3b7ff11276364d2d5523ffc0c/checks/10358505723/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling gimli v0.26.1
   Compiling regex v1.5.6
   Compiling regex-automata v0.1.10
   Compiling object v0.29.0
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/compiler/rustc_data_structures/src/graph/dominators/mod.rs:138:21
   0:     0x7f50838809c6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hafd81a0b5d487c53
   0:     0x7f50838809c6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hafd81a0b5d487c53
   1:     0x7f50838eefb8 - core::fmt::write::h4a778e35e4f38324
   2:     0x7f5083871f81 - std::io::Write::write_fmt::ha216e80cfbe2a850
   3:     0x7f5083880795 - std::sys_common::backtrace::print::h1b9ed32e93c301a6
   4:     0x7f5083883cd7 - std::panicking::default_hook::{{closure}}::h1690eea152a3b836
   5:     0x7f5083883a36 - std::panicking::default_hook::h5e9cc4e411668de1
   6:     0x7f5084252142 - rustc_driver[54a292329d61ee30]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:     0x7f5083884603 - std::panicking::rust_panic_with_hook::h0e85789db00c4c44
   8:     0x7f50838842f2 - std::panicking::begin_panic_handler::{{closure}}::h27551f22330a5e57
   9:     0x7f5083880f6c - std::sys_common::backtrace::__rust_end_short_backtrace::h0c44fb4de016d510
  11:     0x7f5083836123 - core::panicking::panic_fmt::hea4f8765df0a9f41
  11:     0x7f5083836123 - core::panicking::panic_fmt::hea4f8765df0a9f41
  12:     0x7f50838361fd - core::panicking::panic::h6c8c2e19c88621fe
  13:     0x7f5084c93669 - rustc_data_structures[c12f58c5cf5ee10e]::graph::dominators::dominators::<&&rustc_middle[3fee2c4d4e9f0dfe]::mir::basic_blocks::BasicBlocks>
  14:     0x7f5084d0ed54 - <&mut <rustc_mir_transform[95d13cbdfb44c867]::ctfe_limit::CtfeLimit as rustc_middle[3fee2c4d4e9f0dfe]::mir::MirPass>::run_pass::{closure#0} as core[cada5d954255392c]::ops::function::FnMut<((rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlock, &rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlockData),)>>::call_mut
  15:     0x7f5084c4c93f - <core[cada5d954255392c]::iter::adapters::map::Map<core[cada5d954255392c]::iter::adapters::enumerate::Enumerate<core[cada5d954255392c]::slice::iter::Iter<rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlockData>>, <rustc_index[3c4177419627218c]::vec::IndexVec<rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlock, rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlockData>>::iter_enumerated::{closure#0}> as core[cada5d954255392c]::iter::traits::iterator::Iterator>::try_fold::<(), core[cada5d954255392c]::iter::traits::iterator::Iterator::find_map::check<(rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlock, &rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlockData), rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlock, &mut <rustc_mir_transform[95d13cbdfb44c867]::ctfe_limit::CtfeLimit as rustc_middle[3fee2c4d4e9f0dfe]::mir::MirPass>::run_pass::{closure#0}>::{closure#0}, core[cada5d954255392c]::ops::control_flow::ControlFlow<rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlock>>
  16:     0x7f5084d3e9ab - <alloc[7b8a314fc46dcffe]::vec::Vec<rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlock> as alloc[7b8a314fc46dcffe]::vec::spec_from_iter::SpecFromIter<rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlock, core[cada5d954255392c]::iter::adapters::filter_map::FilterMap<core[cada5d954255392c]::iter::adapters::map::Map<core[cada5d954255392c]::iter::adapters::enumerate::Enumerate<core[cada5d954255392c]::slice::iter::Iter<rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlockData>>, <rustc_index[3c4177419627218c]::vec::IndexVec<rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlock, rustc_middle[3fee2c4d4e9f0dfe]::mir::BasicBlockData>>::iter_enumerated::{closure#0}>, <rustc_mir_transform[95d13cbdfb44c867]::ctfe_limit::CtfeLimit as rustc_middle[3fee2c4d4e9f0dfe]::mir::MirPass>::run_pass::{closure#0}>>>::from_iter
  17:     0x7f5084d2e459 - <rustc_mir_transform[95d13cbdfb44c867]::ctfe_limit::CtfeLimit as rustc_middle[3fee2c4d4e9f0dfe]::mir::MirPass>::run_pass
  18:     0x7f5084d9b144 - rustc_mir_transform[95d13cbdfb44c867]::pass_manager::run_passes_inner
  19:     0x7f5084c2c68e - rustc_mir_transform[95d13cbdfb44c867]::run_analysis_to_runtime_passes
  20:     0x7f5084c2bf0a - rustc_mir_transform[95d13cbdfb44c867]::mir_drops_elaborated_and_const_checked
  21:     0x7f5085eeba51 - rustc_query_system[4ff46d1e41d78cab]::query::plumbing::try_execute_query::<rustc_query_impl[bba56a74d3a8740d]::plumbing::QueryCtxt, rustc_query_system[4ff46d1e41d78cab]::query::caches::DefaultCache<rustc_middle[3fee2c4d4e9f0dfe]::ty::WithOptConstParam<rustc_span[4b119e7221467786]::def_id::LocalDefId>, &rustc_data_structures[c12f58c5cf5ee10e]::steal::Steal<rustc_middle[3fee2c4d4e9f0dfe]::mir::Body>>>
  22:     0x7f5086053bc4 - rustc_query_system[4ff46d1e41d78cab]::query::plumbing::get_query::<rustc_query_impl[bba56a74d3a8740d]::queries::mir_drops_elaborated_and_const_checked, rustc_query_impl[bba56a74d3a8740d]::plumbing::QueryCtxt, rustc_middle[3fee2c4d4e9f0dfe]::dep_graph::dep_node::DepKind>
  23:     0x7f5085b68f63 - <rustc_query_impl[bba56a74d3a8740d]::Queries as rustc_middle[3fee2c4d4e9f0dfe]::ty::query::QueryEngine>::mir_drops_elaborated_and_const_checked
  24:     0x7f508438ee04 - <rustc_session[8fecdfce2d84b36e]::session::Session>::time::<(), rustc_interface[700f389ea52edc67]::passes::analysis::{closure#3}>
  25:     0x7f50843aa932 - rustc_interface[700f389ea52edc67]::passes::analysis
  26:     0x7f5085f2f69d - rustc_query_system[4ff46d1e41d78cab]::query::plumbing::try_execute_query::<rustc_query_impl[bba56a74d3a8740d]::plumbing::QueryCtxt, rustc_query_system[4ff46d1e41d78cab]::query::caches::DefaultCache<(), core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>>
  27:     0x7f50860585c9 - rustc_query_system[4ff46d1e41d78cab]::query::plumbing::get_query::<rustc_query_impl[bba56a74d3a8740d]::queries::analysis, rustc_query_impl[bba56a74d3a8740d]::plumbing::QueryCtxt, rustc_middle[3fee2c4d4e9f0dfe]::dep_graph::dep_node::DepKind>
  28:     0x7f5085b5f39a - <rustc_query_impl[bba56a74d3a8740d]::Queries as rustc_middle[3fee2c4d4e9f0dfe]::ty::query::QueryEngine>::analysis
  29:     0x7f50842ac014 - <rustc_interface[700f389ea52edc67]::passes::QueryContext>::enter::<rustc_driver[54a292329d61ee30]::run_compiler::{closure#1}::{closure#2}::{closure#2}, core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>
  30:     0x7f50842b5a1a - <rustc_interface[700f389ea52edc67]::interface::Compiler>::enter::<rustc_driver[54a292329d61ee30]::run_compiler::{closure#1}::{closure#2}, core[cada5d954255392c]::result::Result<core[cada5d954255392c]::option::Option<rustc_interface[700f389ea52edc67]::queries::Linker>, rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>
  31:     0x7f50842537cb - rustc_span[4b119e7221467786]::with_source_map::<core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>, rustc_interface[700f389ea52edc67]::interface::run_compiler<core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>, rustc_driver[54a292329d61ee30]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  32:     0x7f50842aaf8a - <scoped_tls[5706bf77a85fe418]::ScopedKey<rustc_span[4b119e7221467786]::SessionGlobals>>::set::<rustc_interface[700f389ea52edc67]::interface::run_compiler<core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>, rustc_driver[54a292329d61ee30]::run_compiler::{closure#1}>::{closure#0}, core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>
  33:     0x7f5084270fa0 - std[153e7eca4933144d]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[700f389ea52edc67]::util::run_in_thread_pool_with_globals<rustc_interface[700f389ea52edc67]::interface::run_compiler<core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>, rustc_driver[54a292329d61ee30]::run_compiler::{closure#1}>::{closure#0}, core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>
  34:     0x7f50842b6486 - std[153e7eca4933144d]::panicking::try::<core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>, core[cada5d954255392c]::panic::unwind_safe::AssertUnwindSafe<<std[153e7eca4933144d]::thread::Builder>::spawn_unchecked_<rustc_interface[700f389ea52edc67]::util::run_in_thread_pool_with_globals<rustc_interface[700f389ea52edc67]::interface::run_compiler<core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>, rustc_driver[54a292329d61ee30]::run_compiler::{closure#1}>::{closure#0}, core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  35:     0x7f50842639c4 - <<std[153e7eca4933144d]::thread::Builder>::spawn_unchecked_<rustc_interface[700f389ea52edc67]::util::run_in_thread_pool_with_globals<rustc_interface[700f389ea52edc67]::interface::run_compiler<core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>, rustc_driver[54a292329d61ee30]::run_compiler::{closure#1}>::{closure#0}, core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[cada5d954255392c]::result::Result<(), rustc_errors[e6ef93a3d7f5133c]::ErrorGuaranteed>>::{closure#1} as core[cada5d954255392c]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x7f5083891c3e - std::sys::unix::thread::Thread::new::thread_start::h5d9f8d6d1fd775b4
  37:     0x7f5083626b43 - <unknown>
  38:     0x7f50836b8a00 - <unknown>
  39:                0x0 - <unknown>
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.68.0-nightly (33e84e27b 2022-12-30) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=0 -Z unstable-options -Z binary-dep-depinfo -Z unstable-options -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [mir_drops_elaborated_and_const_checked] elaborating drops for `expr::<impl at /cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.102/src/expr.rs:783:1: 783:10>::DUMMY`
#1 [analysis] running analysis passes on this crate
error: could not compile `syn`
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:19:10
```

</details>

RalfJung commented on compiler/rustc_middle/src/mir/syntax.rs:
```diff
@@ -355,6 +358,8 @@ pub enum StatementKind<'tcx> {
     /// This avoids adding a new block and a terminator for simple intrinsics.
     Intrinsic(Box<NonDivergingIntrinsic<'tcx>>),
 
+    ConstEvalCounter,
```
Please add a comment explaining the behavior of this operation.

RalfJung commented on compiler/rustc_const_eval/src/const_eval/machine.rs:
```diff
@@ -561,8 +561,13 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
-    fn before_terminator(ecx: &mut InterpCx<'mir, 'tcx, Self>) -> InterpResult<'tcx> {
-        // The step limit has already been hit in a previous call to `before_terminator`.
+    // Not used here, but used by Miri (see `src/tools/miri/src/machine.rs`).
+    fn before_terminator(_ecx: &mut InterpCx<'mir, 'tcx, Self>) -> InterpResult<'tcx> {
+        Ok(())
+    }
```
This is the default impl, so you can just remove it entirely.

RalfJung commented on compiler/rustc_mir_transform/src/ctfe_limit.rs:
```diff
@@ -0,0 +1,56 @@
+};
+use rustc_middle::ty::TyCtxt;
+
+pub struct CtfeLimit;
```
Please add a doc comment explaining what this does. I first thought it only adds `ConstEvalCounter` for back edges, but it also adds one right before every call, it seems?

bryangarza committed:
Move CtfeLimit MirPass to inner_mir_for_ctfe

bryangarza committed:
During MirBorrowck, ignore ConstEvalCounter

bryangarza committed:
Bless and update consts tests

bryangarza committed:
Bless z-help test for new flag

bryangarza committed:
Add comment on cause of panic in dominators algorithm

bryangarza committed:
Add comments and remove unnecessary code

bryangarza commented:
@rustbot ready

tmiasko commented on compiler/rustc_mir_transform/src/ctfe_limit.rs:
```diff
@@ -0,0 +1,58 @@
+    node_data: &BasicBlockData<'_>,
+) -> bool {
+    let doms = basic_blocks.dominators();
+    basic_blocks.indices().any(|potential_dom| {
+        doms.is_reachable(potential_dom)
+            && doms.is_reachable(node)
+            && doms.is_dominated_by(node, potential_dom)
```
It looks like iterating over `doms.dominators(node)` would reduce the complexity by the number of blocks factor.

bryangarza replied:
Thanks, I was able to improve the code a lot thanks to your comment 😊. Before, I tried to solve it this way because I couldn't figure out how to access the non-immediate dominators of each node, but now I see that the `doms.dominators(node)` method gives us an iterator over that data. I added some doc comments to `dominators/mod.rs` for later reference!

bryangarza committed:
Improve efficiency of has_back_edge(...)

RalfJung commented:
Those were drive-by comments. Thanks for resolving them! The new comments look good.

But I won't be able to do a full review.

albertlarsan68 commented:
Tests have moved from `src/test` to `tests`, please [rebase](https://rustc-dev-guide.rust-lang.org/git.html#rebasing) on top of new `master`.
@rustbot author

bryangarza committed:
Move const-eval/stable-metric ui tests

bryangarza commented:
@rustbot ready

oli-obk commented:
🤦 turns out I already reviewed this entirely two weeks ago

@bors r+

bors commented:
:pushpin: Commit c4c3edd000ac1b68b889619ba423e2b8558231d9 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk c4c3edd000ac1b68b889619ba423e2b8558231d9 -->
<!-- homu: {"type":"Approved","sha":"c4c3edd000ac1b68b889619ba423e2b8558231d9","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #106976) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit bdb815a22ab00450dcc010a99309c24c475432a6 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk bdb815a22ab00450dcc010a99309c24c475432a6 -->
<!-- homu: {"type":"Approved","sha":"bdb815a22ab00450dcc010a99309c24c475432a6","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit bdb815a22ab00450dcc010a99309c24c475432a6 with merge 3cdd0197e78e2ca3403157617709bbfe18a3f0b5...
<!-- homu: {"type":"BuildStarted","head_sha":"bdb815a22ab00450dcc010a99309c24c475432a6","merge_sha":"3cdd0197e78e2ca3403157617709bbfe18a3f0b5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4035057612/jobs/6937627420)
Approved by: oli-obk
Pushing 3cdd0197e78e2ca3403157617709bbfe18a3f0b5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4035057612/jobs/6937627420"},"merge_sha":"3cdd0197e78e2ca3403157617709bbfe18a3f0b5"} -->

rust-timer commented:
Finished benchmarking commit (3cdd0197e78e2ca3403157617709bbfe18a3f0b5): [comparison URL](https://perf.rust-lang.org/compare.html?start=bcb064a7f4aee705c97ce1b74beb9af0f9b5adb2&end=3cdd0197e78e2ca3403157617709bbfe18a3f0b5&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=bcb064a7f4aee705c97ce1b74beb9af0f9b5adb2&end=3cdd0197e78e2ca3403157617709bbfe18a3f0b5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3% | [0.3%, 0.4%] | 3     |
| Regressions ❌ <br /> (secondary)  | 0.8% | [0.5%, 1.0%] | 9     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.3% | [0.3%, 0.4%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bcb064a7f4aee705c97ce1b74beb9af0f9b5adb2&end=3cdd0197e78e2ca3403157617709bbfe18a3f0b5&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.3% | [2.3%, 2.3%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.0% | [1.8%, 2.2%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.3% | [2.3%, 2.3%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bcb064a7f4aee705c97ce1b74beb9af0f9b5adb2&end=3cdd0197e78e2ca3403157617709bbfe18a3f0b5&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-2.4%, -2.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

oli-obk commented on compiler/rustc_mir_transform/src/ctfe_limit.rs:
```diff
@@ -0,0 +1,59 @@
+
+pub struct CtfeLimit;
+
+impl<'tcx> MirPass<'tcx> for CtfeLimit {
+    #[instrument(skip(self, _tcx, body))]
+    fn run_pass(&self, _tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let doms = body.basic_blocks.dominators();
```
The remaining perf regression can likely be fixed by implementing the `is_enabled` method of this trait and gating the pass on `tiny_const_eval_limit` so that the `ConstEvalCounter`s are only added if they are actually used.

oli-obk replied:
Nevermind, I misremembered where the flag is used.

pnkfelix commented:
I agree with @oli-obk 's comment above

@rustbot label: +perf-regression-triaged
